tools/envoy_collect/envoy_collect.py

```diff
@@ -0,0 +1,243 @@
+    manifest = access_log_paths + list(dump_handlers_paths.values()) + [
+        modified_envoy_config_path, envoy_log_path
+    ]
+    # This is where we will find out where the admin endpoint is listening.
+    admin_address_path = os.path.join(envoy_tmpdir, 'admin_address.txt')
+
+    # Only run under 'perf record' in performance mode.
+    if perf:
+      perf_data_path = os.path.join(envoy_tmpdir, 'perf.data')
+      manifest.append(perf_data_path)
+      perf_record_args = [
+          PERF_PATH,
+          'record',
+          '-o',
+          perf_data_path,
+          '-g',
+          '--',
+      ]
+    else:
+      perf_record_args = []
+
+    # This is how we will invoke the wrapped envoy.
+    envoy_shcmd = ' '.join(
+        map(pipes.quote, perf_record_args + [
+            parse_result.envoy_binary,
+            '-c',
+            modified_envoy_config_path,
+            '-l',
+            'error' if perf else 'trace',
+            '--admin-address-path',
+            admin_address_path,
+        ] + unknown_args[1:]))
+    print(envoy_shcmd)
+
+    # Some process setup stuff to ensure the child process gets cleaned up properly if the
+    # collector dies and doesn't get its signals implicity.
+    def EnvoyPreexecFn():
+      os.setpgrp()
+      libc = ctypes.CDLL(ctypes.util.find_library('c'), use_errno=True)
+      libc.prctl(PR_SET_PDEATHSIG, signal.SIGTERM)
+
+    # Launch Envoy, register for SIGINT, and wait for the child process to exit.
+    with open(envoy_log_path, 'w') as envoy_log:
+      envoy_proc = sp.Popen(
+          envoy_shcmd,
+          stdin=sp.PIPE,
+          stderr=envoy_log,
+          preexec_fn=EnvoyPreexecFn,
+          shell=True)
+
+      def SignalHandler(signum, frame):
+        # The read is deferred until the signal so that the Envoy process gets a
+        # chance to write the file out.
+        with open(admin_address_path, 'r') as f:
+          admin_address = 'http://%s' % f.read()
+        # Fetch from the admin endpoint.
+        for handler, path in dump_handlers_paths.items():
+          handler_url = '%s/%s' % (admin_address, handler)
+          print('Fetching %s' % handler_url)
+          with open(path, 'w') as f:
+            f.write(FetchUrl(handler_url))
+        # Send SIGINT to Envoy process, it should exit and execution will
+        # continue from the envoy_proc.wait() below.
+        print('Sending Envoy process (PID=%d) SIGINT...' % envoy_proc.pid)
+        envoy_proc.send_signal(signal.SIGINT)
+
+      signal.signal(signal.SIGINT, SignalHandler)
+      return_code = envoy_proc.wait()
+
+    # Collect manifest files and tar them.
+    with tarfile.TarFile(parse_result.output_path, 'w') as output_tar:
+      for path in manifest:
+        if os.path.exists(path):
+          output_tar.add(path, arcname=os.path.basename(path))
+
+    # Dump the Envoy log to stdout, this is usful when debugging
+    # envoy_collect.py itself.
+    with open(envoy_log_path, 'r') as f:
```

Comment:
I'll drop and take a TODO. I was trying to get this to work with `tee`, but hit some signal and process control issues. Can do this in Python itself, although I'm imagining a situation in which the performance of the Python logging doesn't keep up with trace output from Envoy :)